# Modeling Notebook

Prepare and merge narrative similarity data, alignment structure, ratings, full text, and extracted events.

In [1]:
from pathlib import Path
import json
import pandas as pd
import numpy as np

In [2]:
base = Path.cwd()
data_dir = base / '..' / 'data'

paths = {
    'modeling_overall': data_dir / 'Modeling Data' / 'modeling_data_overall.csv',
    'modeling_structural': data_dir / 'Modeling Data' / 'modeling_data_structural.csv',
    'pairs': data_dir / 'Qualtrics_Story_Pairs_sampled_clean.csv',
    'alignments': data_dir / 'Alignment' / 'sampled_story_alignments.json',
}

for name, p in paths.items():
    print(f'{name}: {p} | exists={p.exists()}')


modeling_overall: /Users/shayan/Projects/NarrativeSimilarity/src/../data/Modeling Data/modeling_data_overall.csv | exists=True
modeling_structural: /Users/shayan/Projects/NarrativeSimilarity/src/../data/Modeling Data/modeling_data_structural.csv | exists=True
pairs: /Users/shayan/Projects/NarrativeSimilarity/src/../data/Qualtrics_Story_Pairs_sampled_clean.csv | exists=True
alignments: /Users/shayan/Projects/NarrativeSimilarity/src/../data/Alignment/sampled_story_alignments.json | exists=True


In [3]:
modeling_overall_df = pd.read_csv(paths['modeling_overall'])
modeling_structural_df = pd.read_csv(paths['modeling_structural'])
pairs_df = pd.read_csv(paths['pairs'])

with open(paths['alignments'], 'r', encoding='utf-8') as f:
    alignments = json.load(f)

print('modeling_overall_df:', modeling_overall_df.shape)
print('modeling_structural_df:', modeling_structural_df.shape)
print('pairs_df:', pairs_df.shape)
print('alignments:', len(alignments))


modeling_overall_df: (374, 18)
modeling_structural_df: (364, 18)
pairs_df: (120, 10)
alignments: 120


In [4]:
alignment_df = pd.DataFrame([
    {
        'PairID': item.get('PairID'),
        'EventsA_align': item.get('EventsA', []),
        'EventsB_align': item.get('EventsB', []),
        'alignment': item.get('alignment', {}),
        'matches': item.get('alignment', {}).get('matches', []),
        'unmatched_a': item.get('alignment', {}).get('unmatched_a', []),
        'unmatched_b': item.get('alignment', {}).get('unmatched_b', []),
        'num_matches': len(item.get('alignment', {}).get('matches', [])),
        'num_unmatched_a': len(item.get('alignment', {}).get('unmatched_a', [])),
        'num_unmatched_b': len(item.get('alignment', {}).get('unmatched_b', [])),
    }
    for item in alignments
])

for df in [modeling_overall_df, modeling_structural_df, pairs_df, alignment_df]:
    df['PairID'] = df['PairID'].astype(str).str.strip()

alignment_df[['PairID', 'num_matches', 'num_unmatched_a', 'num_unmatched_b']].head()

,PairID,num_matches,num_unmatched_a,num_unmatched_b
0,P131,0,4,2
1,P189,0,1,6
2,P000,0,5,3
3,P196,0,3,1
4,P072,0,7,10


In [5]:
pair_meta_cols = ['PairID', 'bucket', 'sim_full', 'sim_event', 'FullA', 'FullB', 'EventsA', 'EventsB']
pair_meta_df = pairs_df[pair_meta_cols].drop_duplicates(subset=['PairID'])

overall_merged_df = modeling_overall_df.merge(alignment_df, on='PairID', how='left').merge(
    pair_meta_df, on='PairID', how='left', suffixes=('', '_pairs')
)

structural_merged_df = modeling_structural_df.merge(alignment_df, on='PairID', how='left').merge(
    pair_meta_df, on='PairID', how='left', suffixes=('', '_pairs')
)

pair_level_merged_df = alignment_df.merge(pair_meta_df, on='PairID', how='left')

print('overall_merged_df:', overall_merged_df.shape)
print('structural_merged_df:', structural_merged_df.shape)
print('pair_level_merged_df:', pair_level_merged_df.shape)


overall_merged_df: (374, 34)
structural_merged_df: (364, 34)
pair_level_merged_df: (120, 17)


In [6]:
print('Core columns available for modeling:')
print([
    'PairID', 'bucket', 'sim_full', 'sim_event',
    'num_matches', 'num_unmatched_a', 'num_unmatched_b',
    'FullA', 'FullB', 'EventsA', 'EventsB',
    'full_rating', 'event_rating'
])

pair_level_merged_df[['PairID', 'bucket', 'sim_full', 'sim_event', 'num_matches', 'num_unmatched_a', 'num_unmatched_b']].head()


Core columns available for modeling:
['PairID', 'bucket', 'sim_full', 'sim_event', 'num_matches', 'num_unmatched_a', 'num_unmatched_b', 'FullA', 'FullB', 'EventsA', 'EventsB', 'full_rating', 'event_rating']


,PairID,bucket,sim_full,sim_event,num_matches,num_unmatched_a,num_unmatched_b
0,P131,HL,0.407780,0.123518,0,4,2
1,P189,HL,0.446355,0.069382,0,1,6
2,P000,HH,0.746363,0.483632,0,5,3
3,P196,HL,0.454492,0.125337,0,3,1
4,P072,HH,0.381345,0.320885,0,7,10


## Alignment Distance: `D_alignment`

This section computes `D_alignment` using the full alignment dictionary and event-level aligned pairs.

- `cost_skip = (E_A - |A_aligned|) + (E_B - |B_aligned|)` where aligned sets come from all match groups.
- `cost_reorder` is inversion count over the induced `B` sequence from aligned event pairs `(i, j)` after sorting by `i`.
- `Z = |A_pairs|(|A_pairs|-1)/2 + (E_A + E_B)`.

So `D_alignment = (cost_reorder + cost_skip) / Z`, and `alignment_similarity = 1 - D_alignment`.

In [7]:
def compute_alignment_distance(row):
    # Uses full alignment dictionary with event-level aligned pairs.
    alignment = row.get('alignment') or {}
    matches = alignment.get('matches', []) or []

    events_a = row.get('EventsA_align') or []
    events_b = row.get('EventsB_align') or []
    E_A = len(events_a)
    E_B = len(events_b)

    aligned_a = set()
    aligned_b = set()
    aligned_pairs = []

    # Build event-level aligned pairs A_pairs via many-to-many expansion.
    for match in matches:
        a_inds = match.get('a_indices', []) or []
        b_inds = match.get('b_indices', []) or []

        aligned_a.update(a_inds)
        aligned_b.update(b_inds)

        for i in a_inds:
            for j in b_inds:
                aligned_pairs.append((i, j))

    cost_skip = (E_A - len(aligned_a)) + (E_B - len(aligned_b))

    # Reordering inversions in B after sorting aligned pairs by A index.
    aligned_pairs.sort(key=lambda x: x[0])
    b_sequence = [j for (_, j) in aligned_pairs]

    cost_reorder = 0
    for i in range(len(b_sequence)):
        for j in range(i + 1, len(b_sequence)):
            if b_sequence[i] > b_sequence[j]:
                cost_reorder += 1

    num_pairs = len(aligned_pairs)
    max_reorder = num_pairs * (num_pairs - 1) / 2
    Z = max_reorder + (E_A + E_B)

    D_alignment = (cost_reorder + cost_skip) / Z if Z > 0 else 0.0

    return pd.Series({
        'cost_reorder': float(cost_reorder),
        'cost_skip': float(cost_skip),
        'D_alignment': float(D_alignment),
        'alignment_similarity': float(1 - D_alignment),
    })

alignment_distance_df = alignment_df.apply(compute_alignment_distance, axis=1)
alignment_df = pd.concat([alignment_df, alignment_distance_df], axis=1)

print(alignment_df[['D_alignment', 'alignment_similarity']].describe().to_string())
alignment_df[['PairID', 'cost_reorder', 'cost_skip', 'D_alignment', 'alignment_similarity']].head()


       D_alignment  alignment_similarity
count   120.000000            120.000000
mean      0.867946              0.132054
std       0.272089              0.272089
min       0.000000              0.000000
25%       1.000000              0.000000
50%       1.000000              0.000000
75%       1.000000              0.000000
max       1.000000              1.000000


,PairID,cost_reorder,cost_skip,D_alignment,alignment_similarity
0,P131,0.0,6.0,1.0,0.0
1,P189,0.0,7.0,1.0,0.0
2,P000,0.0,8.0,1.0,0.0
3,P196,0.0,4.0,1.0,0.0
4,P072,0.0,17.0,1.0,0.0


## Target Construction: `event_rating_num`

In the next code cell, we prepare participant-level modeling data (one row per participant and `PairID`) by combining overall and structural responses with alignment-derived features.

The raw target column is `event_rating` (text labels such as `"3: moderately similar"`). We convert this into a numeric target, `event_rating_num`, by extracting the leading integer (for example, `3`).

`event_rating_num` is the final target column for prediction models.

In [8]:
# Participant-level modeling target: event_rating (one row per participant x PairID)
target_col = 'event_rating'

participant_model_df = pd.concat([overall_merged_df, structural_merged_df], ignore_index=True)

# Convert labels like '3: moderately similar' -> 3
participant_model_df['event_rating_num'] = (
    participant_model_df[target_col]
    .astype(str)
    .str.extract(r'^(\d+)')[0]
    .astype(float)
)

model_feature_cols = [
    'PairID', 'PROLIFIC_PID', 'bucket', 'sim_full', 'sim_event',
    'num_matches', 'num_unmatched_a', 'num_unmatched_b',
    'FullA', 'FullB', 'EventsA', 'EventsB',
    'event_rating', 'event_rating_num'
]
participant_model_df = participant_model_df[model_feature_cols].copy()

print('participant_model_df shape:', participant_model_df.shape)
print('target column:', target_col)
print('event_rating_num value counts:')
print(participant_model_df['event_rating_num'].value_counts(dropna=False).sort_index().to_string())
participant_model_df.head()


participant_model_df shape: (738, 14)
target column: event_rating
event_rating_num value counts:
event_rating_num
1.0    363
2.0    229
3.0    110
4.0     36


,PairID,PROLIFIC_PID,bucket,sim_full,sim_event,num_matches,num_unmatched_a,num_unmatched_b,FullA,FullB,EventsA,EventsB,event_rating,event_rating_num
0,P010,62029a55ef1cb18a1337c61a,HH,0.564017,0.403043,1,5,1,I wanted to be an artist when I was in high sc...,and I suddenly realized something about scienc...,My dad died when I was six — I saw my mom stru...,I realized I had to pick one thing in science ...,3: moderately similar,3.0
1,P321,62029a55ef1cb18a1337c61a,LL,0.252370,0.129338,0,6,6,"five or six bands, you know, 'cause people wou...",something with business. I didn't know exactly...,became the radio station's band — played 80-10...,I woke up at two o'clock in the morning — I pa...,2: slightly similar,2.0
2,P242,62029a55ef1cb18a1337c61a,LH,0.256391,0.463312,4,3,1,"Where were you when you were our age, 22, 23? ...",I actually had no idea that tech was a thing. ...,I was in Chile — I was about to finish college...,A friend graduated and taught himself how to c...,3: moderately similar,3.0
3,P361,62029a55ef1cb18a1337c61a,LL,0.253082,0.104825,0,7,4,So at Alchemy we are taking the concept of che...,where you weren't truly able to talk about emo...,went and worked at Detroit Country Day School ...,worked up the courage to read a poem at an ope...,1: not similar,1.0
4,P292,666f2488909bc0331b8563e6,LH,0.218787,0.400481,3,1,0,"I was already a pretty avid diver , and enjoye...",So going back into the time when you were firs...,came down to Miami — went to law school — prac...,I came out and had a film — I started looking ...,1: not similar,1.0


In [9]:
print('Participant-level rows by PairID (first 10):')
print(participant_model_df.groupby('PairID').size().head(10).to_string())

print('Average event_rating_num by bucket:')
print(participant_model_df.groupby('bucket')['event_rating_num'].mean().sort_index().to_string())


Participant-level rows by PairID (first 10):
PairID
P000     8
P004     9
P010     4
P013    12
P017     9
P020     7
P024    10
P031     4
P033     7
P034     5
Average event_rating_num by bucket:
bucket
HH    2.084158
HL    1.434211
LH    1.933333
LL    1.389937


## Fit Alignment-Only Model With Intercept

Because `D_alignment` is pair-level, we tune an alignment-only linear model against the mean `event_rating_num` per `PairID`:

`event_rating_num_mean ≈ c + alpha * alignment_similarity`

where `c` is an intercept (noise/bias term).

In [10]:
# Rebuild merges so they include D_alignment and full alignment columns.
overall_merged_df = modeling_overall_df.merge(alignment_df, on='PairID', how='left').merge(
    pair_meta_df, on='PairID', how='left', suffixes=('', '_pairs')
)
structural_merged_df = modeling_structural_df.merge(alignment_df, on='PairID', how='left').merge(
    pair_meta_df, on='PairID', how='left', suffixes=('', '_pairs')
)

participant_model_df = pd.concat([overall_merged_df, structural_merged_df], ignore_index=True)
participant_model_df['event_rating_num'] = (
    participant_model_df['event_rating'].astype(str).str.extract(r'^(\d+)')[0].astype(float)
)

pair_target_df = (
    participant_model_df.groupby('PairID', as_index=False)['event_rating_num']
    .mean()
    .rename(columns={'event_rating_num': 'event_rating_num_mean'})
)

pair_model_df = (
    alignment_df[['PairID', 'D_alignment', 'alignment_similarity']]
    .merge(pair_target_df, on='PairID', how='inner')
)

# Keep only rows valid for alignment-only modeling.
valid_alignment = np.isfinite(pair_model_df['alignment_similarity'].astype(float)) & np.isfinite(pair_model_df['event_rating_num_mean'].astype(float))
pair_model_df = pair_model_df.loc[valid_alignment].copy()

# Shared pair-level split used by all models for fair comparison.
rng = np.random.RandomState(42)
all_pair_ids = pair_model_df['PairID'].astype(str).to_numpy()
if len(all_pair_ids) < 13:
    raise ValueError(f'Need at least 13 pairs for test split, found {len(all_pair_ids)}')

test_pair_ids = set(rng.choice(all_pair_ids, size=13, replace=False))
train_pair_ids = set(pid for pid in all_pair_ids if pid not in test_pair_ids)

print(f'Shared split created. Total pairs: {len(all_pair_ids)} | Train: {len(train_pair_ids)} | Test: {len(test_pair_ids)}')
print('First test PairIDs:', sorted(list(test_pair_ids))[:5])



Shared split created. Total pairs: 98 | Train: 85 | Test: 13
First test PairIDs: ['P013', 'P061', 'P097', 'P100', 'P131']


## Model Evaluation: How Well Does Alignment Explain Average Event Rating?

### Method
We evaluate the model at the **story-pair level** because `D_alignment` is pair-level and the target is the mean participant rating per pair (`event_rating_num_mean`).

The fitted model is:

`event_rating_num_mean ≈ alpha * alignment_similarity`, where `alignment_similarity = 1 - D_alignment`.

To assess explanatory power, we report:
- **RMSE** and **MAE** (prediction error in rating units),
- **R²** relative to predicting the global mean,
- **Pearson correlation** between predicted and observed average ratings.

In [11]:
# Alignment-only model evaluated on shared train/test split.
def regression_metrics(df, y_col, pred_col):
    y_true = df[y_col].astype(float)
    y_pred = df[pred_col].astype(float)
    rmse = float(np.sqrt(((y_true - y_pred) ** 2).mean()))
    mae = float((y_true - y_pred).abs().mean())
    ss_res = float(((y_true - y_pred) ** 2).sum())
    ss_tot = float(((y_true - y_true.mean()) ** 2).sum())
    r2 = 1 - (ss_res / ss_tot) if ss_tot > 0 else np.nan
    corr = float(y_true.corr(y_pred)) if len(df) > 1 else np.nan
    return {'RMSE': rmse, 'MAE': mae, 'R2': r2, 'Pearson_r': corr}

alignment_split_df = pair_model_df.copy()
alignment_split_df['PairID'] = alignment_split_df['PairID'].astype(str)

train_alignment_df = alignment_split_df[alignment_split_df['PairID'].isin(train_pair_ids)].copy()
test_alignment_df = alignment_split_df[alignment_split_df['PairID'].isin(test_pair_ids)].copy()

x_train = train_alignment_df['alignment_similarity'].astype(float).to_numpy()
y_train = train_alignment_df['event_rating_num_mean'].astype(float).to_numpy()
X_train = np.column_stack([np.ones(len(x_train)), x_train])
coef_align, *_ = np.linalg.lstsq(X_train, y_train, rcond=None)

c_hat = float(coef_align[0])
alpha_hat = float(coef_align[1])

train_alignment_df['pred_event_rating_mean_from_alignment'] = c_hat + alpha_hat * train_alignment_df['alignment_similarity'].astype(float)
test_alignment_df['pred_event_rating_mean_from_alignment'] = c_hat + alpha_hat * test_alignment_df['alignment_similarity'].astype(float)

train_alignment_metrics = regression_metrics(train_alignment_df, 'event_rating_num_mean', 'pred_event_rating_mean_from_alignment')
test_alignment_metrics = regression_metrics(test_alignment_df, 'event_rating_num_mean', 'pred_event_rating_mean_from_alignment')

print('Alignment-only model (shared split)')
print(f'Train fit: event_rating_num_mean ~= {c_hat:.4f} + {alpha_hat:.4f} * alignment_similarity')
print(f'Train pairs: {len(train_alignment_df)} | Test pairs: {len(test_alignment_df)}')
print('Train metrics:')
print(f"  RMSE: {train_alignment_metrics['RMSE']:.4f}")
print(f"  MAE: {train_alignment_metrics['MAE']:.4f}")
print(f"  R^2: {train_alignment_metrics['R2']:.4f}")
print(f"  Pearson r: {train_alignment_metrics['Pearson_r']:.4f}")
print('')
print('Test metrics:')
print(f"  RMSE: {test_alignment_metrics['RMSE']:.4f}")
print(f"  MAE: {test_alignment_metrics['MAE']:.4f}")
print(f"  R^2: {test_alignment_metrics['R2']:.4f}")
print(f"  Pearson r: {test_alignment_metrics['Pearson_r']:.4f}")

test_alignment_df[['PairID', 'event_rating_num_mean', 'pred_event_rating_mean_from_alignment', 'alignment_similarity']].head(10)



Alignment-only model (shared split)
Train fit: event_rating_num_mean ~= 1.6248 + 1.0397 * alignment_similarity
Train pairs: 85 | Test pairs: 13
Train metrics:
  RMSE: 0.4268
  MAE: 0.3526
  R^2: 0.3116
  Pearson r: 0.5582

Test metrics:
  RMSE: 0.3208
  MAE: 0.2500
  R^2: 0.2781
  Pearson r: 0.7119


,PairID,event_rating_num_mean,pred_event_rating_mean_from_alignment,alignment_similarity
0,P131,1.333333,1.624802,0.000000
10,P237,1.375000,1.624802,0.000000
18,P097,1.500000,1.624802,0.000000
31,P184,1.666667,1.624802,0.000000
40,P210,2.000000,2.326617,0.675000
42,P395,1.666667,1.624802,0.000000
47,P230,2.000000,2.465757,0.808824
62,P365,1.000000,1.624802,0.000000
64,P100,1.571429,1.624802,0.000000
75,P224,2.375000,2.303807,0.653061


## Interpretation: Alignment-Only Model (With Intercept)

- Fitted equation: `event_rating_num_mean ≈ 1.6057 + 1.0122 * alignment_similarity`.
- `RMSE = 0.4136` and `MAE = 0.3367`: errors are substantially lower than the no-intercept version.
- `R^2 = 0.3192`: this model now explains about 31.9% of pair-level variance (better than mean baseline).
- `Pearson r = 0.5649`: moderate positive co-variation remains between predicted and observed ratings.

Adding the intercept materially improved calibration and overall predictive performance for the alignment-only model.

## Modeling Progress

Semantic similarity is now included below; compare the joint-model metrics against the alignment-only model to assess incremental value.

## Semantic Distance

The semantic component evaluates how similar aligned events are in meaning. Let `d(·,·)` denote a normalized semantic distance (e.g., cosine distance between contextual embeddings), bounded in `[0,1]`. We define:


## $ D_{\text{semantic}}(S_a,S_b) = \frac{1}{|\mathcal{A}|} \sum_{(i,j)\in\mathcal{A}} d(e_{ai}, e_{bj}) $

This term captures *what* happens in the narrative: aligned events that are semantically close reduce the overall distance, while conceptually divergent episodes increase it.

In [12]:
from sentence_transformers import SentenceTransformer

st_model = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')

def get_event_text(events, idx):
    if 1 <= idx <= len(events):
        return events[idx - 1]
    if 0 <= idx < len(events):
        return events[idx]
    return None

def cosine_distance_01(vec_a, vec_b):
    denom = (np.linalg.norm(vec_a) * np.linalg.norm(vec_b))
    if denom == 0:
        return 1.0
    cos_sim = float(np.dot(vec_a, vec_b) / denom)
    return (1.0 - cos_sim) / 2.0

def compute_semantic_distance(row):
    alignment = row.get('alignment') or {}
    matches = alignment.get('matches', []) or []
    events_a = row.get('EventsA_align') or []
    events_b = row.get('EventsB_align') or []
    aligned_pairs = []
    for match in matches:
        a_inds = match.get('a_indices', []) or []
        b_inds = match.get('b_indices', []) or []
        for i in a_inds:
            for j in b_inds:
                aligned_pairs.append((i, j))

    if len(aligned_pairs) == 0:
        return pd.Series({'D_semantic': 1.0, 'semantic_similarity': 0.0, 'num_aligned_pairs': 0})

    texts_a, texts_b = [], []
    for i, j in aligned_pairs:
        ta = get_event_text(events_a, i)
        tb = get_event_text(events_b, j)
        if ta is None or tb is None:
            continue
        texts_a.append(ta)
        texts_b.append(tb)

    if len(texts_a) == 0:
        return pd.Series({'D_semantic': 1.0, 'semantic_similarity': 0.0, 'num_aligned_pairs': 0})

    emb_a = st_model.encode(texts_a, convert_to_numpy=True, normalize_embeddings=False)
    emb_b = st_model.encode(texts_b, convert_to_numpy=True, normalize_embeddings=False)
    dists = [cosine_distance_01(a, b) for a, b in zip(emb_a, emb_b)]
    D_semantic = float(np.mean(dists))
    return pd.Series({'D_semantic': D_semantic, 'semantic_similarity': float(1.0 - D_semantic), 'num_aligned_pairs': int(len(dists))})

semantic_df = alignment_df.apply(compute_semantic_distance, axis=1)
alignment_df = pd.concat([alignment_df, semantic_df], axis=1)

pair_target_df = (participant_model_df.groupby('PairID', as_index=False)['event_rating_num']
    .mean().rename(columns={'event_rating_num': 'event_rating_num_mean'}))

pair_sem_model_df = alignment_df[['PairID', 'alignment_similarity', 'semantic_similarity']].merge(pair_target_df, on='PairID', how='inner')
pair_sem_model_df['PairID'] = pair_sem_model_df['PairID'].astype(str)

valid_joint = np.isfinite(pair_sem_model_df[['alignment_similarity', 'semantic_similarity']].to_numpy(dtype=float)).all(axis=1) & np.isfinite(pair_sem_model_df['event_rating_num_mean'].to_numpy(dtype=float))
pair_sem_model_df = pair_sem_model_df.loc[valid_joint].copy()

# Use the same pair split as alignment-only model.
train_joint_df = pair_sem_model_df[pair_sem_model_df['PairID'].isin(train_pair_ids)].copy()
test_joint_df = pair_sem_model_df[pair_sem_model_df['PairID'].isin(test_pair_ids)].copy()

X_train_joint = train_joint_df[['alignment_similarity', 'semantic_similarity']].to_numpy(dtype=float)
y_train_joint = train_joint_df['event_rating_num_mean'].to_numpy(dtype=float)
X_train_joint_aug = np.column_stack([np.ones(len(X_train_joint)), X_train_joint])
coef_joint, *_ = np.linalg.lstsq(X_train_joint_aug, y_train_joint, rcond=None)

c_hat_joint = float(coef_joint[0])
alpha_hat_joint = float(coef_joint[1])
beta_hat_joint = float(coef_joint[2])

train_joint_df['pred_event_rating_mean_joint'] = X_train_joint_aug @ coef_joint
X_test_joint = test_joint_df[['alignment_similarity', 'semantic_similarity']].to_numpy(dtype=float)
X_test_joint_aug = np.column_stack([np.ones(len(X_test_joint)), X_test_joint])
test_joint_df['pred_event_rating_mean_joint'] = X_test_joint_aug @ coef_joint

train_joint_metrics = regression_metrics(train_joint_df, 'event_rating_num_mean', 'pred_event_rating_mean_joint')
test_joint_metrics = regression_metrics(test_joint_df, 'event_rating_num_mean', 'pred_event_rating_mean_joint')

print('Joint model (with intercept, shared split): event_rating_num_mean ≈ c + alpha*alignment_similarity + beta*semantic_similarity')
print(f'c_hat_joint: {c_hat_joint:.4f}')
print(f'alpha_hat_joint: {alpha_hat_joint:.4f}')
print(f'beta_hat_joint: {beta_hat_joint:.4f}')
print(f'Train pairs: {len(train_joint_df)} | Test pairs: {len(test_joint_df)}')
print('Train metrics:')
print(f"  RMSE: {train_joint_metrics['RMSE']:.4f}")
print(f"  MAE: {train_joint_metrics['MAE']:.4f}")
print(f"  R^2: {train_joint_metrics['R2']:.4f}")
print(f"  Pearson r: {train_joint_metrics['Pearson_r']:.4f}")
print('')
print('Test metrics:')
print(f"  RMSE: {test_joint_metrics['RMSE']:.4f}")
print(f"  MAE: {test_joint_metrics['MAE']:.4f}")
print(f"  R^2: {test_joint_metrics['R2']:.4f}")
print(f"  Pearson r: {test_joint_metrics['Pearson_r']:.4f}")
pair_sem_model_df[['PairID', 'event_rating_num_mean', 'alignment_similarity', 'semantic_similarity']].head()



Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


Joint model (with intercept, shared split): event_rating_num_mean ≈ c + alpha*alignment_similarity + beta*semantic_similarity
c_hat_joint: 1.5830
alpha_hat_joint: 0.2021
beta_hat_joint: 0.9339
Train pairs: 85 | Test pairs: 13
Train metrics:
  RMSE: 0.4018
  MAE: 0.3367
  R^2: 0.3898
  Pearson r: 0.6244

Test metrics:
  RMSE: 0.2893
  MAE: 0.2282
  R^2: 0.4128
  Pearson r: 0.7298


,PairID,event_rating_num_mean,alignment_similarity,semantic_similarity
0,P131,1.333333,0.00000,0.000000
1,P000,2.375000,0.00000,0.000000
2,P072,2.000000,0.00000,0.000000
3,P212,2.142857,0.00000,0.000000
4,P292,2.300000,0.96875,0.632487


## Interpretation: Joint Model (With Intercept)

- Fitted equation: `event_rating_num_mean ≈ 1.5672 + 0.1376 * alignment_similarity + 0.9832 * semantic_similarity`.
- `RMSE = 0.3883`, `MAE = 0.3208`, `R^2 = 0.3999`, `Pearson r = 0.6324`.
- Compared to alignment-only with intercept (`RMSE 0.4136`, `R^2 0.3192`), the joint model improves fit and explained variance.
- `beta_hat_joint` is larger than `alpha_hat_joint`, suggesting semantic similarity contributes more than alignment similarity in this linear specification.

Overall, adding semantic similarity provides a clear incremental gain over structure-only prediction.

## Baseline Hypothesis: Full-Story Embeddings Only

To test whether your alignment + semantic features are truly helpful, we add a text-only baseline:

- Use only full story text (`FullA`, `FullB`) for each `PairID`.
- Embed the full stories with a sentence embedding model.
- Train a simple regression head to predict pair-level `event_rating_num_mean`.
- Compare this baseline against the alignment+semantic joint model on the same 85/13 train-test setup.

If the text-only baseline underperforms on test metrics, that supports the hypothesis that explicit alignment and semantic similarity add real value.


In [13]:
# Text-only baseline: full-story embeddings -> regression head for event_rating_num_mean
# Uses the SAME shared train/test PairID split as alignment-only and joint models.

story_df = pairs_df[['PairID', 'FullA', 'FullB']].drop_duplicates(subset=['PairID']).copy()
story_df['PairID'] = story_df['PairID'].astype(str).str.strip()

if 'pair_target_df' not in globals():
    tmp_df = pd.concat([overall_merged_df, structural_merged_df], ignore_index=True)
    tmp_df['event_rating_num'] = tmp_df['event_rating'].astype(str).str.extract(r'^(\d+)')[0].astype(float)
    pair_target_df = (
        tmp_df.groupby('PairID', as_index=False)['event_rating_num']
        .mean()
        .rename(columns={'event_rating_num': 'event_rating_num_mean'})
    )

target_df = pair_target_df.copy()
target_df['PairID'] = target_df['PairID'].astype(str).str.strip()

story_df = story_df.merge(target_df, on='PairID', how='inner')
story_df = story_df.dropna(subset=['FullA', 'FullB', 'event_rating_num_mean']).copy()

# Restrict to the shared split IDs so all models are evaluated on identical pairs.
train_story_df = story_df[story_df['PairID'].isin(train_pair_ids)].copy()
test_story_df = story_df[story_df['PairID'].isin(test_pair_ids)].copy()

if len(train_story_df) == 0 or len(test_story_df) == 0:
    raise ValueError('Shared split produced empty train/test for text-only baseline.')

if 'st_model' not in globals():
    from sentence_transformers import SentenceTransformer
    st_model = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')

train_a = st_model.encode(train_story_df['FullA'].tolist(), convert_to_numpy=True, normalize_embeddings=True)
train_b = st_model.encode(train_story_df['FullB'].tolist(), convert_to_numpy=True, normalize_embeddings=True)
test_a = st_model.encode(test_story_df['FullA'].tolist(), convert_to_numpy=True, normalize_embeddings=True)
test_b = st_model.encode(test_story_df['FullB'].tolist(), convert_to_numpy=True, normalize_embeddings=True)

def make_features(a, b):
    cos = np.sum(a * b, axis=1, keepdims=True)
    return np.hstack([a, b, np.abs(a - b), a * b, cos])

X_train = make_features(train_a, train_b)
X_test = make_features(test_a, test_b)
y_train = train_story_df['event_rating_num_mean'].astype(float).to_numpy()

X_train_aug = np.column_stack([np.ones(len(X_train)), X_train])
coef_story, *_ = np.linalg.lstsq(X_train_aug, y_train, rcond=None)

def predict_linear(X, coef):
    X_aug = np.column_stack([np.ones(len(X)), X])
    return X_aug @ coef

train_story_df['pred_story_baseline'] = predict_linear(X_train, coef_story)
test_story_df['pred_story_baseline'] = predict_linear(X_test, coef_story)

story_train_metrics = regression_metrics(train_story_df, 'event_rating_num_mean', 'pred_story_baseline')
story_test_metrics = regression_metrics(test_story_df, 'event_rating_num_mean', 'pred_story_baseline')

print('Text-only full-story embedding baseline (shared split)')
print(f'Train pairs: {len(train_story_df)} | Test pairs: {len(test_story_df)}')
print('Train metrics:')
print(f"  RMSE: {story_train_metrics['RMSE']:.4f}")
print(f"  MAE: {story_train_metrics['MAE']:.4f}")
print(f"  R^2: {story_train_metrics['R2']:.4f}")
print(f"  Pearson r: {story_train_metrics['Pearson_r']:.4f}")
print('')
print('Test metrics:')
print(f"  RMSE: {story_test_metrics['RMSE']:.4f}")
print(f"  MAE: {story_test_metrics['MAE']:.4f}")
print(f"  R^2: {story_test_metrics['R2']:.4f}")
print(f"  Pearson r: {story_test_metrics['Pearson_r']:.4f}")

print('')
print('=== Shared-Test Comparison (All Models) ===')
if 'test_alignment_metrics' in globals():
    print(f"Alignment-only  | RMSE={test_alignment_metrics['RMSE']:.4f} | MAE={test_alignment_metrics['MAE']:.4f} | R^2={test_alignment_metrics['R2']:.4f} | r={test_alignment_metrics['Pearson_r']:.4f}")
if 'test_joint_metrics' in globals():
    print(f"Alignment+Semantic (Joint) | RMSE={test_joint_metrics['RMSE']:.4f} | MAE={test_joint_metrics['MAE']:.4f} | R^2={test_joint_metrics['R2']:.4f} | r={test_joint_metrics['Pearson_r']:.4f}")
print(f"Full-story text baseline   | RMSE={story_test_metrics['RMSE']:.4f} | MAE={story_test_metrics['MAE']:.4f} | R^2={story_test_metrics['R2']:.4f} | r={story_test_metrics['Pearson_r']:.4f}")

if 'test_alignment_metrics' in globals() and 'test_joint_metrics' in globals():
    print('')
    print('Delta vs text baseline (negative RMSE/MAE and positive R^2/r are better):')
    print(f"  Alignment-only delta: RMSE={test_alignment_metrics['RMSE'] - story_test_metrics['RMSE']:+.4f}, MAE={test_alignment_metrics['MAE'] - story_test_metrics['MAE']:+.4f}, R^2={test_alignment_metrics['R2'] - story_test_metrics['R2']:+.4f}, r={test_alignment_metrics['Pearson_r'] - story_test_metrics['Pearson_r']:+.4f}")
    print(f"  Joint-model delta:    RMSE={test_joint_metrics['RMSE'] - story_test_metrics['RMSE']:+.4f}, MAE={test_joint_metrics['MAE'] - story_test_metrics['MAE']:+.4f}, R^2={test_joint_metrics['R2'] - story_test_metrics['R2']:+.4f}, r={test_joint_metrics['Pearson_r'] - story_test_metrics['Pearson_r']:+.4f}")

test_story_df[['PairID', 'event_rating_num_mean', 'pred_story_baseline']].head(13)



Text-only full-story embedding baseline (shared split)
Train pairs: 85 | Test pairs: 13
Train metrics:
  RMSE: 0.0000
  MAE: 0.0000
  R^2: 1.0000
  Pearson r: 1.0000

Test metrics:
  RMSE: 0.5578
  MAE: 0.4722
  R^2: -1.1825
  Pearson r: -0.0947

=== Shared-Test Comparison (All Models) ===
Alignment-only  | RMSE=0.3208 | MAE=0.2500 | R^2=0.2781 | r=0.7119
Alignment+Semantic (Joint) | RMSE=0.2893 | MAE=0.2282 | R^2=0.4128 | r=0.7298
Full-story text baseline   | RMSE=0.5578 | MAE=0.4722 | R^2=-1.1825 | r=-0.0947

Delta vs text baseline (negative RMSE/MAE and positive R^2/r are better):
  Alignment-only delta: RMSE=-0.2370, MAE=-0.2222, R^2=+1.4606, r=+0.8066
  Joint-model delta:    RMSE=-0.2685, MAE=-0.2440, R^2=+1.5953, r=+0.8245


,PairID,event_rating_num_mean,pred_story_baseline
0,P131,1.333333,1.600926
10,P237,1.375000,1.227565
18,P097,1.500000,2.064681
31,P184,1.666667,2.534767
40,P210,2.000000,1.728216
42,P395,1.666667,1.241852
47,P230,2.000000,1.427053
62,P365,1.000000,1.483326
64,P100,1.571429,2.172039
75,P224,2.375000,1.439175


In [14]:
# Validate script model reproduces notebook joint-model test performance exactly.
from structural_similarity_model import StructuralSimilarityModel, StructuralModelParams

script_model = StructuralSimilarityModel(
    params=StructuralModelParams(
        c_hat_joint=float(c_hat_joint),
        alpha_hat_joint=float(alpha_hat_joint),
        beta_hat_joint=float(beta_hat_joint),
    )
)

alignment_by_pairid = {
    str(row['PairID']): row
    for row in alignment_df[['PairID', 'alignment', 'EventsA_align', 'EventsB_align']].to_dict(orient='records')
}

script_rows = []
for pid in test_joint_df['PairID'].astype(str).tolist():
    pair_row = alignment_by_pairid[pid]
    out = script_model.predict_similarity(pair_row)
    script_rows.append({
        'PairID': pid,
        'pred_event_rating_mean_joint_script': out['pred_event_rating_mean_joint'],
        'alignment_similarity_script': out['alignment_similarity'],
        'semantic_similarity_script': out['semantic_similarity'],
    })

script_test_df = test_joint_df[['PairID', 'event_rating_num_mean', 'pred_event_rating_mean_joint', 'alignment_similarity', 'semantic_similarity']].copy()
script_test_df['PairID'] = script_test_df['PairID'].astype(str)
script_test_df = script_test_df.merge(pd.DataFrame(script_rows), on='PairID', how='inner')

# Compare to 3 decimal places.
np.testing.assert_allclose(
    script_test_df['pred_event_rating_mean_joint_script'].to_numpy(dtype=float),
    script_test_df['pred_event_rating_mean_joint'].to_numpy(dtype=float),
    rtol=0.0,
    atol=1e-3,
)
np.testing.assert_allclose(
    script_test_df['alignment_similarity_script'].to_numpy(dtype=float),
    script_test_df['alignment_similarity'].to_numpy(dtype=float),
    rtol=0.0,
    atol=1e-3,
)
np.testing.assert_allclose(
    script_test_df['semantic_similarity_script'].to_numpy(dtype=float),
    script_test_df['semantic_similarity'].to_numpy(dtype=float),
    rtol=0.0,
    atol=1e-3,
)

script_test_metrics = regression_metrics(
    script_test_df.rename(columns={'pred_event_rating_mean_joint_script': 'y_pred'}),
    'event_rating_num_mean',
    'y_pred',
)

for k in ['RMSE', 'MAE', 'R2', 'Pearson_r']:
    np.testing.assert_allclose(script_test_metrics[k], test_joint_metrics[k], rtol=0.0, atol=1e-3)

print('Script model test predictions and metrics match notebook joint model to 3 decimal places.')
print(script_test_metrics)

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


Script model test predictions and metrics match notebook joint model to 3 decimal places.
{'RMSE': 0.28932829612422223, 'MAE': 0.22815967188617955, 'R2': 0.4128227332809449, 'Pearson_r': 0.7298094626995547}
